### ISSUE TITLE
Document data migration options

### ISSUE THREAD

**Issue by ninamak** [18F/OPRE-Unicorn#34]
Document options for db migration and impacts that options might have on other tech choices (a lot of documentation already lives in GH)

Figure out if the AWS Data Migration Service is compatible with cloud.gov-brokered AWS

Determine where we will have the db(s), at least for v0.1, and what factors influence the v1 decision.

**Notes from Milestone doc**
* Milestone 1 will have db in cloud.gov 
* Milestone 2+ will work on the main part of the db migration path
* Factors for v0.1 vs v1:
  * ATO implications
  * If we can get the data into cloud.gov, that would get us a lot of benefits. We can inherit more of their security/ATO stuff. Also some services we get to benefit from
  * Does ACF AWS environment provide awesome services?
  * How big is the db? (Would need GT to provide access to staging/prod db to determine.) How much of a lift to move to cloud.gov?
  * How much permissions will ACF OCIO give us? (Will they allow VPC peering? Configurations?)
  * Script to copy data from AWS to cloud.gov db? (cloud.gov not that mature of a product, there are probably dragons)

**Tasks**

- [ ]

**Comment by amymok** [comment]
Some thoughts I have on Data Migration Options:

The current database is an Oracle database in vendor’s AWS, we need to move the database under ACF environment and convert it to an  open source database (possibly Postgres Database, and I will use Postgres throughout this comment).  This work will include:
- Schema Migration
- Schema Transformation
- Data Migration
- Data Transformation
- Data Syncing

Couple strategies we are thinking about:

1. Using AWS Database Migration Service:
- This service allows all the tasks to happen easily.
- Dependencies:
    - Current vendor needs to open up a public access connection to current source database
    - Current vendor grants access for us to access the source database to review what is in the database for verification during/after migration
    - Target database needs to exist in AWS
    - Access to ACF/HHS AWS, where we can do the following but not limited to:
        - Create and open connection to database and replication instances
        - Create EC2 client
        - Run Database Migration Services
        - Run Schema Conversion Tool
        - Set up data sync
- Couple variations we can do:
  - Vendor’s Oracle -> ACF/HHS’s Oracle -> ACF/HHS’s Postgres
      - Pros:
          - Allows us to move vendor’s Oracle database to ACF environment as soon as possible without the need to figure out how to fix any migration issues cause by different types of databases.
          - Direct copying from one Oracle database to another, which probably means we don’t need too much direct access to the vendor’s Oracle database to resolve issue; we can go to the ACF Oracle database to analyze issues.
          - If there’s a need to set up scripts or other triggers when migrating and syncing, we can do it from ACF environment
      - Cons:
          - Means it is doing two migrations, which means it probably cost OPRE more money to keep two different databases around, and two migration services and syncing being run. 
  - Vendor’s Oracle -> ACF/HHS’s Postgres
      - Pros:
          - Allows the least amount of resources within AWS, may cost a bit less to have just one conversion and sync.
      - Cons:
          - ACF does not have the full copy of the current database.
          - May not have as much access and permission to vendor’s database, which may mean more time coordinations are needed with vendor to get different things from them

2. Using ora2pg and other open source conversion tools:

- Dependencies:
    - Current vendor needs to open up a public access connection to current source database
    - Current vendor grants permission for us to access the source database to review what is in the database for verification during/after migration
    - Need a space big enough to hold all the scripts for schema and data to load into Postgres Database when running the conversion
- Open questions:
    - If we run the conversion tools on our local machine, will that be sufficient (memory/space)?  If not, where should this happen?
    - Where should we put this database, ACF/HHS AWS, cloud.gov?  How do we plan to load the data in?
    - Unclear how to handle data syncing at this point with this strategy, would we need to build and run scripts to look at vendor’s database periodically to get the latest data and load into the Postgres database?

Another piece we need to figure out is what role of the Postgres Database played after it is converted.
- Is this just for reference?
- Are we picking and choosing what’s needed to incorporate to the new UNICORN as we develop?  If so, does that mean this new database exists in the AWS or cloud.gov?  Another migration?  Or manually picking and choosing?
- Are we using this new converted database as-is in UNICORN?  Would that mean we need to develop associated application models for this purpose?
- If the converted database exists in AWS, any restrictions that prevent us to use it as the [user-provided service instance](https://docs.cloudfoundry.org/devguide/services/user-provided.html) in the cloud.gov?
    - Do we need to do so?
- If we build a new database on cloud.gov, picking and choosing what we need as we go, do we need to periodically do data sync with this converted database?

**Comment by ninamak** [comment]
@amymok @carjug  Can we also verify that the tools/services you want to use will be compatible with the Oracle12 DB (even though it will no longer be supported by AWS after 3/22)?

**Comment by carjug** [comment]
@amymok Everything in your comment makes sense to me.

The only thing I have to add to is what was put in the original issue:

> "Figure out if the AWS Data Migration Service is compatible with cloud.gov-brokered AWS"

In our discussions with Mark Headd, when we talked about using the AWS data migration service, it sounded like going from the vendor's AWS Oracle -> HHS/AWS AWS Postgres and then having the cloud.gov application connect to the AWS Postgres DB would be the simplest path for us. Apparently it would be a pretty big lift to integrate a cloud.gov db with the AWS Migration Service directly. So, just so we're all on the same page: in the case that we use the AWS migration service for the database migration, we would never need to touch a cloud.gov db instance. We would only use a cloud.gov db instance if we made the decision to then migrate the AWS Postgres DB over to cloud.gov after the initial migration.

**Comment by amymok** [comment]
@carjug Thanks for calling that out.  Yes, that's correct.  I probably was doing it too implicitly under the last part:
>Another piece we need to figure out is what role of the Postgres Database played after it is converted.

We would need to decide what we want to do with the database we moved over, do we want to start a new set of schema and database for Unicorn vs, can we even use the database as-is.  We would need to decide that path: a new database separated from the copy/conversion of the vendor database in AWS, or that new database exists in cloud.gov.

The question is also what are we integrating the migrated database on AWS with?  Updating pointers to MAPS1.0?  Or more?

**Comment by amymok** [comment]
Data migration will be a big effort.  There are still a lot of unknowns and tasks that involve with this work.  We should perform a few tasks to help us understand the scale and need for this effort:

- ### Path 1: Using AWS Database Migration Service
  - **_Request access and permission to ACF/HHS AWS_**
    - Identify services we need to use for database migration from AWS documentation
    - Initiate the talk with the right personnel to get access
    - Should this task happens before the next task?  Or should we be more informed by doing the next task first?  There's also a chance if we won't get access then the next task is irrelevant.
  - **_Perform [test runs](https://docs.aws.amazon.com/dms/latest/userguide/CHAP_GettingStarted.html) with a sample database_**
    - Test run homogeneous migration with sample database (Oracle to Oracle)
    - Test run heterogeneous migration with sample database (Oracle to PostgreSQL)
    - **Purpose**:
      - Identify the steps and effort it takes to do the migration by practicing it
      - Identify the AWS services we need to use
      - Identify more targeted asks on getting ACF/HHS AWS permission, and add to our previous request
      - Identify what we need from the vendor
      - Identify the potential issues we may run into when migrating and perform early mitigation
    - **How can we do this?**
      - We may be able to perform this under TTS sandbox, we would need to find out if we can, if so, we can start early
      - If we cannot perform this under TTS sandbox, we can do it in ACF/HHS AWS environment
    - This task should help us scope out the actual work
  - **_Request access and permission to vendor's Oracle test database for database migration_**
    - Previous task should help us understand and detail out this task
    - This task will be a test run for test database for us to understand if there's anything else we need to mitigate
    - Once we identify additional needs, we will also need to request similar access and permission to vendor's staging and production Oracle databases
- ### Path 2: Using ora2pg and other open source tool
  - It would be best if Path 1 works, if not, we will need to scope this out then
- ### Cloud.gov piece
  - **_Spin up an application to test out user-provided service instance_**
    - This task depends on a few things:
      - After we know the tech stack so we can mimic the environment
      - After having an AWS PostgreSQL database

These task should help us scope out later tasks and identify needs to change course if the original path does not work out.

@carjug, I do have a few questions I would like to discuss so we can be on the same page in terms on what new databases we need in different environments.  I started [this mural](https://app.mural.co/t/gsa6/m/gsa6/1625005563169/ef52441e1f7328f8c9880d912f1cfbe8324b3898) so we can collaborate on this discussion.

**Comment by carjug** [comment]
@amymok the issues and tasks look good to me! the only thing I might add is that maybe we can try all of the db migration stuff in the tts sandbox first, before we make our request to acf ocio, because that may allow us more information about what exactly we'll be doing and what we'll need from them.

**Comment by amymok** [comment]
Yes, I am actually more comfortable with because then we will be more informed.  I think I put the ask at the beginning just in case we will not get access, we don't need to work on the AWS tasks since they are no longer relevant.

With this change, the ordering for the tasks in Path 1 is as follow:
- Perform test runs with a sample database within the same VPC in AWS
- Perform test runs with a sample database ~~or MAPS development database~~ across different VPCs
- Initiate access and permission request to ACF/HHS AWS (We may even ask for VPC Peering at this point, TBD)
- Perform test runs with MAPS development database within the same VPC in ACF/HHS AWS
- Update access and permission request to ACF/HHS AWS (Anything we found out if we need to update the request)
- Perform test runs with ~~a sample database or~~ MAPS development database across different VPCs
- Request access and permission to vendor's Oracle development database for database migration
- Perform database migration from vendor's MAPS development database to AHS/HHS AWS's Oracle and PostgreSQL databases
- Perform database migration from vendor's MAPS staging and production database to AHS/HHS AWS's Oracle and PostgreSQL databases

**Comment by ninamak** [comment]
@amymok Looks like the next step would be to turn these notes and decisions into a rough map of the different phases and create associated cards.  Want to meet Fri afternoon to start doing this?

**Comment by amymok** [comment]
@ninamak Sure!

**Comment by ninamak** [comment]
Update: Amy and I met to turn [bullets 1 & 2](https://github.com/18F/OPRE-Unicorn/issues/34#issuecomment-872506593) into cards, which I'll clean up before Planning on Monday.  

We did not create cards beyond that because 1) they're blocked by needing to work with OCIO and 2) the details will be informed by what gets learned in the earlier work.

We also discussed copying the notes from here on Strategies and Paths into a living Google Doc that we can update as we learn things.  @amymok I'm passing back to you for that.

**Comment by ninamak** [comment]
Update: I created #71, #74 , #75 , and #76 to capture the tasks associated with test runs using a sample db

**Comment by amymok** [comment]
I added the [Database Migration Planning doc](https://docs.google.com/document/d/1xDtV_M7NbVY4LisNOZ40bUN5IOpoJ4pvL7nB3lAb-bo/edit#), which has the above plan.  @ninamak Let me know what you think.  We can probably close this issue and do the planning in the document.  We can refer to the document as we add more cards.

**Comment by ninamak** [comment]
Looks great, thank you!
